rllib/env/multi_agent_env_runner.py

```diff
@@ -0,0 +1,887 @@
+
+        while ts < num_timesteps:
+            # Act randomly.
+            if random_actions:
+                # Note, to get sampled actions from all agents' action
+                # spaces we need to call `MultiAgentEnv.action_space_sample()`.
+                if self.env.unwrapped._action_space_iin_preferred_format:
+                    actions = self.env.action_space.sample()
+                # Otherwise, `action_space_sample()` needs to be implemented.
+                else:
+                    actions = self.env.action_space_sample()
+                # Remove all actions for agents that had no observation.
+                actions = {
+                    agent_id: agent_action
+                    for agent_id, agent_action in actions.items()
+                    if agent_id in obs
+                }
+
+            else:
+                # TODO (simon): This is not correct `forward()` expects
+                # `SampleBatchType`.
+                # Note, `RLModule`'s `forward()` methods expect `NestedDict`s.
+                # Note, we only consider for states and obs agents that step.
+                batch: MultiAgentDict = {
+                    self._agent_id_to_module_id[agent_id]: {
+                        STATE_IN: tree.map_structure(
+                            lambda s: self._convert_from_numpy(s),
+                            states[self._agent_id_to_module_id[agent_id]],
+                        ),
+                        SampleBatch.OBS: self._convert_from_numpy(
+                            np.expand_dims(agent_obs, axis=0)
+                        ),
+                    }
+                    for agent_id, agent_obs in obs.items()
+                }
+                # TODO (Sven, Simon): The `RLModule` has `SampleBatchType` as input
+                # type. Only the _forward_x()` methods have a `NestedDict`. Shall we
+                # compile to `SampleBatchType` here and in `SingleAgentEnvRunner`?
+                from ray.rllib.utils.nested_dict import NestedDict
+
+                batch = NestedDict(batch)
+
+                # Explore or not.
+                if explore:
+                    fwd_out = self.marl_module.forward_exploration(batch)
+                else:
+                    fwd_out = self.marl_module.forward_inference(batch)
+
+                # Sample the actions or draw randomly.
+                actions, action_logps = self._sample_actions_if_necessary(
+                    fwd_out,
+                    explore=explore,
+                )
+
+                # Convert to numpy for recording later to the episode.
+                fwd_out = tree.map_structure(convert_to_numpy, fwd_out)
+
+                # Assign the new states for the agents that stepped.
+                if STATE_OUT in fwd_out:
+                    states.update(tree.map_structure(lambda s: s[STATE_OUT], fwd_out))
+
+            obs, rewards, terminateds, truncateds, infos = self.env.step(actions)
+
+            env_steps += 1
```

Comment:
Ah, cool, you are already doing this! Perfect. Forget most of the above comments, then, except for the docstring ones :)